synapse/state/__init__.py

```diff
@@ -268,6 +268,11 @@ async def compute_event_context(
             The event context.
         """
 
+        logger.info(
+            "compute_event_context is_outlier=%s event_id=%s",
+            event.internal_metadata.is_outlier(),
+            event.event_id,
+        )
         if event.internal_metadata.is_outlier():
```

Comment:
What's the thinking behind `outlier` not being part of the `internal_metadata_dict `?

`outlier` is part of `internal_metadata` but when we serialize it various places, we just use [`internal_metadata.get_dict()`](https://github.com/matrix-org/synapse/blob/e2b8a90897e137fd118768a3bf35b70642916eb7/synapse/events/builder.py#L170) and `outlier` does not get passed on.

Reply:
> What's the thinking behind outlier not being part of the internal_metadata_dict like the rest of the properties there? synapse/events/__init__.py#L113

I've not followed the rest of this PR, but as the person who changed this recently:

We have a separate `outlier` column. I think this is largely historical, but also means that we can select against it without jumping through too many server-side hoops. Given that is the case, we should avoid also storing it in the serialised json, where it wastes space.

consider `outlier` in the same way as `stream_ordering`: it's stored in the `EventInternalMetadata` class, but not in the json.

Reply:
Thanks @richvdh! I think this means I have to adjust the event `builder` and the `EventBase` for how it constructs `_EventInternalMetadata` and marks the resultant event as an `outlier` as well.

Currently, we just pass on the JSON dict, [`synapse/events/builder.py#L269`](https://github.com/matrix-org/synapse/blob/0d87c6bd121d043d5474cc20e352794ff8cf5625/synapse/events/builder.py#L269)

Reply:
Context where `outlier` was moved out of the dict, https://github.com/matrix-org/synapse/pull/9411